### PR TITLE
Ap

### PR DESCRIPTION
**Pull request by tinapai**
ap things and beginnings of tests

### CODE REVIEW

File: core/src/main/java/com/pipai/adv/backend/battle/engine/BattleBackend.kt [hunk 1]
```diff
@@ -52,6 +56,51 @@ class BattleBackend(private val npcList: NpcList, private val battleMap: BattleM
...
+class NoActionIfNotEnoughApRule : CommandRule {
+    override fun canBeExecuted(command: BattleCommand, state: BattleState, unitPositions: Map<Int, GridPosition>): ExecutableStatus {
+        if (command is ActionCommand) {
+            if (!hasEnoughAp(command.unitId, command.requiredAp, state.apState)) {
+                return ExecutableStatus(false, "Not enough action points available")
+            }
+        }
+        return ExecutableStatus.COMMAND_OK
+    }
+    fun hasEnoughAp(npcId: Int, requiredAp: Int, apState: ActionPointState): Boolean {
```
Comment (cypai) [inline core/src/main/java/com/pipai/adv/backend/battle/engine/BattleBackend.kt]: Should be private

File: core/src/main/java/com/pipai/adv/backend/battle/engine/BattleBackend.kt [hunk 2]
```diff
@@ -62,15 +111,43 @@ data class ExecutableStatus(val executable: Boolean, val reason: String?) {
...
                        val npcList: NpcList,
                        val battleMap: BattleMap,
-                       val battleLog: BattleLog)
+                       val battleLog: BattleLog,
+                       val apState: ActionPointState)
+
+class ActionPointState(npcList: NpcList) {
+    val startingNumAPs = 2
+    private var apMap: MutableMap<Int, Int> = HashMap(npcList.associateBy({it.key}, {startingNumAPs})) //todo: find out if this is not efficient
+
```
Comment (cypai) [inline core/src/main/java/com/pipai/adv/backend/battle/engine/BattleBackend.kt]: Add a function that returns if the npcId exists

File: core/src/test/java/com/pipai/adv/backend/battle/domain/ApTest.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+import com.pipai.adv.backend.battle.engine.MoveCommand;
+import com.pipai.adv.npc.Npc;
+import com.pipai.adv.npc.NpcList;
+import com.pipai.test.fixtures.TestFixturesKt;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.util.Arrays;
+
+public class ApTest {
```
Comment (cypai) [inline core/src/test/java/com/pipai/adv/backend/battle/domain/ApTest.java]: Should be ActionPointStateTest

File: core/src/test/java/com/pipai/adv/backend/battle/domain/ApTest.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        //Npc npc = TestFixturesKt.npcFromStats(new UnitStats(1, 1, 1, 1, 1, 1, 1, 1, 3));
+        //int id = npcList.addNpc(npc);
+        //ActionPointState aps = new ActionPointState(npcList);
+        //each npc should start with 2 aps
+        //aps.increaseAp(1, 1)
+        //Assert.assertTrue(aps.getNpcAp(1)==2)
+        //also test the null npc can't increase ap
+    }
+    @Test
+    public void testCantMoveMoreThanApAllow() {
```
Comment (cypai) [inline core/src/test/java/com/pipai/adv/backend/battle/domain/ApTest.java]: Should be moved out to the MoveBackendTest

File: core/src/test/java/com/pipai/adv/backend/battle/domain/ApTest.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+        BattleBackend backend = new BattleBackend(npcList, map);
+
+        MoveCommand cmd = new MoveCommand(1, Arrays.asList(new GridPosition(2, 1), new GridPosition(3, 1)));
+
+        Assert.assertTrue(backend.canBeExecuted(cmd).getExecutable());
+
+        backend.execute(cmd);
+        backend.execute(cmd);
+        Assert.assertFalse(backend.canBeExecuted(cmd).getExecutable());
```
Comment (cypai) [inline core/src/test/java/com/pipai/adv/backend/battle/domain/ApTest.java]: Please also test for the reason it is not executable
